src/PixelGraph.cpp

```diff
@@ -509,4 +509,337 @@ int PixelGraph::calculateValence(cv::Point p) {
+        // ending on the top
+        cv::Vec3b topColor = img->at<cv::Vec3b>(cv::Point(p.x, p.y - 1));
+        cv::Vec3b rightColor = img->at<cv::Vec3b>(cv::Point(p.x + 1, p.y));
+        cv::Vec3b leftColor = img->at<cv::Vec3b>(cv::Point(p.x - 1, p.y));
+
+        m.at<cv::Vec3b>(cv::Point(0, 0)) = leftColor;
+        m.at<cv::Vec3b>(cv::Point(0, 1)) = leftColor;
+        m.at<cv::Vec3b>(cv::Point(1, 0)) = leftColor;
+        m.at<cv::Vec3b>(cv::Point(2, 0)) = rightColor;
+        m.at<cv::Vec3b>(cv::Point(3, 0)) = rightColor;
+        m.at<cv::Vec3b>(cv::Point(3, 1)) = rightColor;
+
+    }
+    else if (n->right != NULL) {
+        // ending on the left
+        cv::Vec3b topColor = img->at<cv::Vec3b>(cv::Point(p.x, p.y - 1));
+        cv::Vec3b bottomColor = img->at<cv::Vec3b>(cv::Point(p.x, p.y + 1));
+        cv::Vec3b leftColor = img->at<cv::Vec3b>(cv::Point(p.x - 1, p.y));
+
+        m.at<cv::Vec3b>(cv::Point(1, 0)) = topColor;
+        m.at<cv::Vec3b>(cv::Point(0, 0)) = topColor;
+        m.at<cv::Vec3b>(cv::Point(0, 1)) = topColor;
+        m.at<cv::Vec3b>(cv::Point(0, 2)) = bottomColor;
+        m.at<cv::Vec3b>(cv::Point(0, 3)) = bottomColor;
+        m.at<cv::Vec3b>(cv::Point(1, 3)) = bottomColor;
+
+    }
+    else if (n->left != NULL) {
+        // ending on the right
+        cv::Vec3b topColor = img->at<cv::Vec3b>(cv::Point(p.x, p.y - 1));
+        cv::Vec3b rightColor = img->at<cv::Vec3b>(cv::Point(p.x + 1, p.y));
+        cv::Vec3b bottomColor = img->at<cv::Vec3b>(cv::Point(p.x, p.y + 1));
+
+        m.at<cv::Vec3b>(cv::Point(2, 0)) = topColor;
+        m.at<cv::Vec3b>(cv::Point(3, 0)) = topColor;
+        m.at<cv::Vec3b>(cv::Point(3, 1)) = topColor;
+        m.at<cv::Vec3b>(cv::Point(3, 2)) = bottomColor;
+        m.at<cv::Vec3b>(cv::Point(3, 3)) = bottomColor;
+        m.at<cv::Vec3b>(cv::Point(2, 3)) = bottomColor;
+    }
+}
+
+void PixelGraph::renderDiagEndPixel(cv::Mat& m, cv::Point p) {
+    // flood with colour first, and then figure out which corners to change
+    cv::Vec3b color = img->at<cv::Vec3b>(p);
+    for (int y = 0; y < 4; y++) {
+        for (int x = 0; x < 4; x++) {
+            m.at<cv::Vec3b>(cv::Point(x, y)) = color;
+        }
+    }
+
+    Node* n = graph->at(p.y)->at(p.x);
+
+    cv::Vec3b topColor = img->at<cv::Vec3b>(cv::Point(p.x, p.y - 1));
+    cv::Vec3b rightColor = img->at<cv::Vec3b>(cv::Point(p.x + 1, p.y));
+    cv::Vec3b bottomColor = img->at<cv::Vec3b>(cv::Point(p.x, p.y + 1));
+    cv::Vec3b leftColor = img->at<cv::Vec3b>(cv::Point(p.x - 1, p.y));
+
+    if (n->topLeft != NULL || n->bottomRight != NULL) {
+        // this diagonal goes \  //
+        m.at<cv::Vec3b>(cv::Point(2, 0)) = topColor;
+        m.at<cv::Vec3b>(cv::Point(3, 0)) = topColor;
+        m.at<cv::Vec3b>(cv::Point(3, 1)) = topColor;
+        m.at<cv::Vec3b>(cv::Point(0, 3)) = leftColor;
+        m.at<cv::Vec3b>(cv::Point(1, 3)) = leftColor;
+        m.at<cv::Vec3b>(cv::Point(0, 2)) = leftColor;

```

Comment:
Only `topColor` and `leftColor` here, but then in the next case we only use `leftColor` and `rightColor`? Isn't this inconsistent and rather asymmetrical? Should `n->topLeft != NULL` and `n->bottomRight != NULL` really be handled in the same case anyway?

Reply:
So this currently results in an end pixel that looks like
```
x x - -      - - x x
x x x -  or  - x x x
- x x x      x x x - 
- - x x      x x - -
```
so the cases are essentially the same. I actually just went with what looked correct in the most cases visually here, but I think that there could be an argument for a slightly more sophisticated method of choosing which colours to take. The big issue that I noticed most often when trying to do this is that there are a lot of cases where you really DON'T want the bottom colour, since it's an end point itself, and it starts looking really weird. II'll try out using left and right in the \ case as well